datafusion/expr/src/udf.rs

```diff
@@ -124,22 +169,116 @@ impl ScalarUDF {
         &self.aliases
     }
 
-    /// Returns this function's signature (what input types are accepted)
+    /// Returns this function's [`Signature`] (what input types are accepted)
     pub fn signature(&self) -> &Signature {
         &self.signature
     }
 
-    /// Return the type of the function given its input types
+    /// The datatype this function returns given the input argument input types
     pub fn return_type(&self, args: &[DataType]) -> Result<DataType> {
         // Old API returns an Arc of the datatype for some reason
         let res = (self.return_type)(args)?;
         Ok(res.as_ref().clone())
     }
 
-    /// Return the actual implementation
+    /// Return an [`Arc`] to the function implementation
     pub fn fun(&self) -> ScalarFunctionImplementation {
         self.fun.clone()
     }
+}
+
+impl<F> From<F> for ScalarUDF
+where
+    F: ScalarUDFImpl + Send + Sync + 'static,
+{
+    fn from(fun: F) -> Self {
+        Self::new_from_trait(fun)
+    }
+}
+
+/// Trait for implementing [`ScalarUDF`].
```

Comment:
Here is the proposed new trait. I think we can use this trait to add things such as  "pre-compiling" arguments https://github.com/apache/arrow-datafusion/issues/8051 and adding better examples / documentation add examples and description to scalar/aggregate functions  #8366.

cc @universalmind303 for your comments

Reply:
I think this looks much more intuitive than the current implementation. I actually just commented on an open issue about the api before reviewing this & my suggestion was nearly identical!

https://github.com/apache/arrow-datafusion/issues/8568#issuecomment-1866966567